### PR TITLE
[Wave9] Progress Bar UI

### PR DESCRIPTION
**Pull request by MaciejSWM**
https://linear.app/expensify-swm/issue/EXP-274/[wave-9]-new-onboarding-flow-progress-bar-[just-the-ui]

UI and Storybook for the Progress bar. To test, easiest way is to `npm run storybook`
http://localhost:6006/?path=/story/components-headerwithbackbutton--progress-bar

Storybook seems to be broken after merge. Looks like it's broken on main as well. In this case best way to test is to do `progressBarPercentage = 25,` in `src/components/HeaderWithBackButton/index.tsx` and to open workspace switcher or any screen with header

### CODE REVIEW

File: src/components/HeaderWithBackButton/types.ts [hunk 1]
```diff
@@ -114,6 +114,12 @@ type HeaderWithBackButtonProps = Partial<ChildrenProps> & {
 
     /** Whether we should overlay the 3 dots menu */
     shouldOverlayDots?: boolean;
+
+    /** Whether we should show progress bar */
+    shouldShowProgressBar: boolean;
+
+    /** 0 - 100 number indicating current progress of the progress bar */
+    progressBarPercentage: number;
```
Comment (kowczarz) [inline src/components/HeaderWithBackButton/types.ts:122]: These are non-null props, but you're setting the default value [here](https://github.com/software-mansion-labs/expensify-app-fork/pull/64/files#diff-3ea703b86598917b6d2bd0c62d7489450ad5d73328aedb4596694419fcccc77eR59-R60), not sure what was the purpose, but that's something that needs to be changed.

**Review by kowczarz — COMMENTED** [review]
One minor code issue, and run `typecheck` because there are some issues in files you didn't change, but they are affected by your changes.

File: src/components/HeaderWithBackButton/types.ts (hunk 1 above)
Comment (MaciejSWM) [reply src/components/HeaderWithBackButton/types.ts:122]: + @mateuuszzzzz we will drop pair-props and stay with just one prop https://docs.google.com/document/d/1sxG_RV5HTFHgdCsJUDNpqSGs09W_OPLBKbNYQfu3FHw/edit#heading=h.qcbt20goffpx sorry for a late change

File: src/components/HeaderWithBackButton/index.tsx [hunk 2]
```diff
@@ -70,6 +71,33 @@ function HeaderWithBackButton({
     // If the icon is present, the header bar should be taller and use different font.
     const isCentralPaneSettings = !!icon;
 
+    let middleContent = null;
```
Comment (kowczarz) [inline src/components/HeaderWithBackButton/index.tsx:74]:
```suggestion
    let middleContent;
```

File: src/components/HeaderWithBackButton/index.tsx [hunk 3]
```diff
@@ -70,6 +71,33 @@ function HeaderWithBackButton({
...
+        );
+    } else {
+        middleContent = (
+            <Header
+                title={title}
+                subtitle={stepCounter ? translate('stepCounter', stepCounter) : subtitle}
+                textStyles={titleColor ? [StyleUtils.getTextColorStyle(titleColor)] : []}
+            />
+        );
+    }
```
Comment (kowczarz) [inline src/components/HeaderWithBackButton/index.tsx:99]: I would consider wrapping it in `useMemo`

File: src/components/HeaderWithBackButton/types.ts [hunk 4]
```diff
@@ -121,6 +121,9 @@ type HeaderWithBackButtonProps = Partial<ChildrenProps> & {
 
     /** Whether we should overlay the 3 dots menu */
     shouldOverlayDots?: boolean;
+
+    /** 0 - 100 number indicating current progress of the progress bar */
+    progressBarPercentage?: number;
```
Comment (kowczarz) [inline src/components/HeaderWithBackButton/types.ts:126]: Hmm, how about adding a type that allow you to define precise range like:
```TS
type Range<T extends number> = number extends T ? number :_Range<T, []>;
type _Range<T extends number, R extends unknown[]> = R['length'] extends T ? R['length'] : R['length'] | _Range<T, [T, ...R]>;
```
and then 
```TS
progressBarPercentage?: Range<100>;
```
Cc. TS guru @blazejkustra

Comment (blazejkustra) [reply src/components/HeaderWithBackButton/types.ts:126]: HAHAHA @kowczarz thanks for the title 💘 

I think it's a little too over-engineered, simple number seems alright here

**Review by blazejkustra — APPROVED** [review]

**Pull request merged**